### PR TITLE
kubeadm: introduce the UnversionedKubeletConfigMap feature gate

### PR DESCRIPTION
**Pull request by neolit123**
#### What this PR does / why we need it:

introduce a new kubeadm specific FG called UnversionedKubeletConfigMap.
this feature controls the old vs new formatting of the kubelet config map used for storing the KubeletConfiguration.

also:
- adapt unit tests and e2e tests
- remove unused function DownloadConfig in kubelet.go

#### Which issue(s) this PR fixes:

xref https://github.com/kubernetes/kubeadm/issues/1582

#### Does this PR introduce a user-facing change?

```release-note
kubeadm: add the kubeadm specific, Alpha (disabled by default) feature gate UnversionedKubeletConfigMap. When this feature is enabled kubeadm will start using a new naming format for the ConfigMap where it stores the KubeletConfiguration structure. The old format included the Kubernetes version - "kube-system/kubelet-config-1.22", while the new format does not - "kube-system/kubelet-config". A similar formatting change is done for the related RBAC rules. The old format is now DEPRECATED and will be removed after the feature graduates to GA. When writing the ConfigMap kubeadm (init, upgrade apply) will respect the value of UnversionedKubeletConfigMap, while when reading it (join, reset, upgrade), it would attempt to use new format first and fallback to the legacy format if needed.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/2915-kubeadm-replace-kubelet-config-x.y
- [Enhancement tracking issue]: https://github.com/kubernetes/enhancements/issues/2915
```

### CODE REVIEW

**Comment by pacoxu** [comment]
The legacy bool works well but looks a little odd to me. The PR LGTM.
Let me re-think the implementation if there are some other ways and get back to review this PR tommorrow. 😄

**Review by pacoxu — APPROVED** [review]
/lgtm

**Review by fabriziopandini — COMMENTED** [review]
Some questions about the upgrade path for existing clusters.
- users not expressing any preference on ConfigMap naming; legacy name will be converted to new unified name during the first upgrade after the default value for the feature flag has been switched from false to true. Is that right? Is the old configMap/RBAC rule being deleted or not?
- users that wants to opt in in the new naming schema early; they have to force the feature gate to true in the kubeadm-config ConfigMap before an upgrade, no matter of the the default value for the feature flag. Is that right?
- users that wants to stick on the legacy schema naming ConfigMap; in oder to prevent kubeadm moving to the new naming schema automatically, they have to force the feature gate to false in the kubeadm-config ConfigMap before the first upgrade after the default value for the feature flag has been switched from false to true. Is that right?

Can we have tests for those scenarios? currently all the upgrade logic is implicit in kubeletConfigFromCluster, which is really nested in the call of chains and it is kind of hard to validate at first sight..

File: cmd/kubeadm/app/componentconfigs/configset_test.go [hunk 1]
```diff
@@ -26,17 +26,26 @@ import (
...
+	}
 	return &kubeadmapi.ClusterConfiguration{
 		KubernetesVersion: constants.CurrentKubernetesVersion.String(),
+		FeatureGates:      map[string]bool{features.UnversionedKubeletConfigMap: true},
 	}
 }
 
 func TestDefault(t *testing.T) {
-	clusterCfg := testClusterCfg()
+	clusterCfg := testClusterCfg(false)
```
[the remark this answers is not in the document]
Comment (neolit123) [reply cmd/kubeadm/app/componentconfigs/configset_test.go:48]: makes sense.

**Comment by neolit123** [comment]
the summary for the current state of the patch:
- during writes (init, post-upgrade) kubeadm will choose the CM/RBAC names based on the FG.
- during reads (join, reset, pre-upgrade) kubeadm will first try the new CM naming format and if it fails it will fallback to the old naming independent of FG
- kubeadm does not read the RBACs, but if users read them they need to adapt manually.

> users not expressing any preference on ConfigMap naming; legacy name will be converted to new unified name during the first upgrade after the default value for the feature flag has been switched from false to true. Is that right? Is the old configMap/RBAC rule being deleted or not?

currently kubeadm does not remove old CM/RBACs during upgrade.
so an old cluster will have old the ...x.yy objects.
cleanup of the old names was stated as "maybe" in the KEP.

> users that wants to opt in in the new naming schema early; they have to force the feature gate to true in the kubeadm-config ConfigMap before an upgrade, no matter of the the default value for the feature flag. Is that right?

if the user sets the FG to true in the ClusterConfiguration before upgrade, kubeadm will try to read the new naming and fallback to the old naming (this is the current state of the patch). after upgrade kubeadm will write the relevant CM in the new location.

> users that wants to stick on the legacy schema naming ConfigMap; in oder to prevent kubeadm moving to the new naming schema automatically, they have to force the feature gate to false in the kubeadm-config ConfigMap before the first upgrade after the default value for the feature flag has been switched from false to true. Is that right?

yes, if the FG is forced to false, during upgrade kubeadm will write to the old location, during join, reset, etc kubeadm will try to use the new name and if it fails it will fallback to the old name.

> Can we have tests for those scenarios? currently all the upgrade logic is implicit in kubeletConfigFromCluster, which is really nested in the call of chains and it is kind of hard to validate at first sight..

my eval is that testing the behavior in unit tests with the current state of the upgrade code is difficult.
i'm looking at testing the following with e2e tests during alpha:
- set the FG to true during init and check if the right CM is created
- create a cluster without the FG, patch the FG to true, upgrade to the same version, make sure that new CM is created

**Comment by fabriziopandini** [comment]
> kubeadm will first try the new CM naming format and if it fails it will fallback to the old naming independent of FG 

@neolit123 I'm ok with this approach given that, as you confirmed above, we are covering all the use cases.
I also understand that alternatives to this will require very invasive changes we don't want to do given that this is a temporary and hopefully short transition.

> my eval is that testing the behavior in unit tests with the current state of the upgrade code is difficult.
i'm looking at testing the following with e2e tests during alpha (1.23)

how dow you think to proceed about implementing E2E testing? merge this PR and the implements E2E? does this means that we are considering a rollback if E2E does not provide the expected signal before release?

**Comment by neolit123** [comment]
Yes. I plan to write the e2e as soon as this merges. Already did local
testing without kinder.

Will update the PR today.

**Comment by neolit123** [comment]
made the changes requested in https://github.com/kubernetes/kubernetes/pull/105741#pullrequestreview-796292161

**Pull request merged**